File: ios/Sources/MultandoSDK/Services/BlockchainService.swift
import Foundation

/// Blockchain token and staking operations.
public final class BlockchainService: Sendable {

    private let httpClient: HTTPClient

    init(httpClient: HTTPClient) {
        self.httpClient = httpClient
    }

    /// Get the token balance for the authenticated user.
    public func balance() async throws -> TokenBalance {
        try await httpClient.request(
            method: "GET",
            path: "/api/v1/blockchain/balance"
        )
    }

    /// Stake tokens.
    public func stake(amount: Double) async throws -> StakingInfo {
        try await httpClient.request(
            method: "POST",
            path: "/api/v1/blockchain/stake",
            body: StakeRequest(amount: amount)
        )
    }

    /// Unstake tokens.
    public func unstake(amount: Double) async throws -> StakingInfo {
        try await httpClient.request(
            method: "POST",
            path: "/api/v1/blockchain/unstake",
            body: StakeRequest(amount: amount)
        )
    }

    /// Get current staking information.
    public func stakingInfo() async throws -> StakingInfo {
        try await httpClient.request(
            method: "GET",
            path: "/api/v1/blockchain/staking-info"
        )
    }

    /// List token transactions.
    public func transactions(page: Int = 1, pageSize: Int = 20) async throws -> [TokenTransaction] {
        try await httpClient.request(
            method: "GET",
            path: "/api/v1/blockchain/transactions",
            queryItems: [
                URLQueryItem(name: "page", value: String(page)),
                URLQueryItem(name: "page_size", value: String(pageSize))
            ]
        )
    }

    /// Claim accumulated staking rewards.
    public func claimRewards() async throws -> ClaimRewardsResponse {
        try await httpClient.request(
            method: "POST",
            path: "/api/v1/blockchain/claim-rewards"
        )
    }
}

File: android/multando-sdk/src/main/kotlin/com/multando/sdk/services/BlockchainService.kt
package com.multando.sdk.services

import com.multando.sdk.core.HttpClient
import com.multando.sdk.models.*

/**
 * Blockchain token and staking operations.
 */
class BlockchainService internal constructor(
    private val httpClient: HttpClient
) {

    /** Get the token balance for the authenticated user. */
    suspend fun balance(): TokenBalance =
        httpClient.request(method = "GET", path = "/api/v1/blockchain/balance")

    /** Stake tokens. */
    suspend fun stake(amount: Double): StakingInfo =
        httpClient.request(
            method = "POST",
            path = "/api/v1/blockchain/stake",
            body = StakeRequest(amount)
        )

    /** Unstake tokens. */
    suspend fun unstake(amount: Double): StakingInfo =
        httpClient.request(
            method = "POST",
            path = "/api/v1/blockchain/unstake",
            body = StakeRequest(amount)
        )

    /** Get current staking information. */
    suspend fun stakingInfo(): StakingInfo =
        httpClient.request(method = "GET", path = "/api/v1/blockchain/staking-info")

    /** List token transactions. */
    suspend fun transactions(page: Int = 1, pageSize: Int = 20): List<TokenTransaction> =
        httpClient.request(
            method = "GET",
            path = "/api/v1/blockchain/transactions",
            queryParams = mapOf(
                "page" to page.toString(),
                "page_size" to pageSize.toString()
            )
        )

    /** Claim accumulated staking rewards. */
    suspend fun claimRewards(): ClaimRewardsResponse =
        httpClient.request(method = "POST", path = "/api/v1/blockchain/claim-rewards")
}
